src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -2,12 +2,162 @@
+                    {
+                        throw new HPackEncodingException();
+                    }
+                    return false;
+                }
+
+                length += headerLength;
+            } while (_enumerator.MoveNext());
+
+            return true;
+        }
+
+        private int EncodeStatusCode(int statusCode, Span<byte> buffer)
+        {
+            switch (statusCode)
+            {
+                case 200:
+                case 204:
+                case 206:
+                case 304:
+                case 400:
+                case 404:
+                case 500:
+                    buffer[0] = (byte)(0x80 | StaticTable.StatusIndex[statusCode]);
+                    return 1;
+                default:
+                    // Send as Literal Header Field Without Indexing - Indexed Name
+                    buffer[0] = 0x08;
+
+                    var statusBytes = StatusCodes.ToStatusBytes(statusCode);
+                    buffer[1] = (byte)statusBytes.Length;
+                    ((Span<byte>)statusBytes).CopyTo(buffer.Slice(2));
+
+                    return 2 + statusBytes.Length;
+            }
+        }
+
+        private bool EncodeHeader(string name, string value, Span<byte> buffer, out int length)
+        {
+            var i = 0;
+            length = 0;
+
+            if (buffer.Length == 0)
+            {
+                return false;
+            }
+
+            buffer[i++] = 0;
+
+            if (i == buffer.Length)
+            {
+                return false;
+            }
+
+            if (!EncodeString(name, buffer.Slice(i), out var nameLength, lowercase: true))
+            {
+                return false;
+            }
+
+            i += nameLength;
+
+            if (i >= buffer.Length)
+            {
+                return false;
+            }
+
+            if (!EncodeString(value, buffer.Slice(i), out var valueLength, lowercase: false))
+            {
+                return false;
+            }
+
+            i += valueLength;
+
+            length = i;
+            return true;
+        }
+
+        private bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        {
+            const int toLowerMask = 0x20;
+
+            var i = 0;
+            length = 0;
+
+            if (buffer.Length == 0)
+            {
+                return false;
+            }
+
+            buffer[0] = 0;
+
+            if (!IntegerEncoder.Encode(s.Length, 7, buffer, out var nameLength))
+            {
+                return false;
+            }
+
+            i += nameLength;
+
+            // TODO: use huffman encoding
+            for (var j = 0; j < s.Length; j++)
+            {
+                if (i >= buffer.Length)
+                {
+                    return false;
+                }
+
+                buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= (byte)'A' && s[j] <= (byte)'Z' ? toLowerMask : 0));
+            }
```

Comment:
Maybe I'm remembering incorrectly, but I thought we already tuned this code in corefx to, for example, remove the unnecessary branching that's occurring here on each byte.

Reply:
Fundamentally this is why we are doing this merge. We had no idea this code path was tuned. What optimization was done in CoreFx? Is it in static bool EncodeStringLiteral?

Reply:
It was.  I had to look back through history to see what I was remembering.  We first changed EncodeStringLiteral to move the branch to no longer be per iteration:
https://github.com/dotnet/corefx/pull/35324/files#diff-ddec2c87b42f404061f4cd18008cfcc7R200-R218
and then subsequently removed the branching entirely:
https://github.com/dotnet/corefx/commit/85cec01822bc70fbc45a25001997b0c1e71b1d22#diff-ddec2c87b42f404061f4cd18008cfcc7R200

Reply:
Can you clarify why this removes branch? There is still a ternary in 
```c#
destination[i] = (byte)((uint)(c - 'A') <= ('Z' - 'A') ? c | 0x20 : c);
```

Reply:
Also, this code is only used by Kestrel, which we will optimize/synchronize in a follow-up PR.

Reply:
I did the first optimization: https://github.com/dotnet/corefx/pull/35324/files#diff-ddec2c87b42f404061f4cd18008cfcc7R200-R218. The second change is a bit more work, which I'd rather do when generally merging the HPackEncoder implementations.

Reply:
> Can you clarify why this removes branch? There is still a ternary

I didn't say it removes all branches.  In the code in this PR:
```C#
buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= (byte)'A' && s[j] <= (byte)'Z' ? toLowerMask : 0));
```
that line is inside the inner loop, which means every iteration of the loop is checking lowercase and if it's false then or'ing 0 with s[j]; that check can instead be hoisted out, so if lowercase is false, there's nothing to be done on each iteration than to just copy from one buffer to the other.  Then there's the `s[j] >= (byte)'A' && s[j] <= (byte)'Z'` which ideally the JIT would be able to optimize to eliminate the branch for the logical and, but currently it can't, so changing it to `(uint)(c - 'A') <= ('Z' - 'A')` removes that branch.